doc/developer/design/20230110_window_functions.md

```diff
@@ -0,0 +1,331 @@
+# Efficient Window Functions
+
+By “window functions”, this document means the `OVER` clause, e.g.,
+`SELECT row_number() OVER (PARTITION BY col1 ORDER BY col2) FROM relation;`
+
+**(Note that [temporal windows](https://materialize.com/docs/sql/patterns/temporal-filters/) are entirely different from what we discuss in this doc. Our support for those is already quite fine. There is no need to use SQL’s `OVER` clause for those.)**
+
+# Overview
+
+[Many users want to use window functions](https://www.notion.so/Window-Functions-Use-Cases-6ad1846a7da942dc8fa28997d9c220dd), but our current window function support is very inefficient: We recompute results for an entire window partition for any small change in the partition. So the only situations when our current support works is if the window partitions are either very small, or they rarely change.
+
+## Goals
+
+We would like to have efficient window function support.
+
+Some window functions are impossible to efficiently support in streaming, because sometimes small input changes cause big result changes. (E.g., if a new first element of a partition appears, then ROW_NUMBERs will change for the whole window partition.) So a realistic goal would be to support at least those cases where a small input change leads to a small output change.
+
+- LAG/LEAD (i.e., previous/next element of the window partition)
+    - We aim for only offset 1 in the first version, which is the default. Bigger offsets not seen in user queries yet, but shouldn’t be a problem to add support later.
```

Comment:
"shouldn’t be a problem to add support later." -> Please elaborate, it's not clear why that is, or why we don't do it now.

Reply:
Moving this to the implementation section.